### ISSUE TITLE
Enable IPv6 DHCP by default

### ISSUE THREAD

**Issue by zfLQ2qx2** [mcrute/alpine-ec2-ami#70]
I was able to build the AMIs with no issue. If I bring up an instance in a vpc with ipv6 enabled I get both IPv4 and IPv6 addresses assigned in the ec2 console, I'm able to ssh into the IPv4 address, however it looks to me like no IPv6 address is being obtained nor routes set at the OS level. I do see the ipv6 module loaded and it is in /etc/modules as well but even following the Alpine wiki I'm not able to obtain an IPv6 address. Anyone tell me what steps I've missed?

**Comment by mcrute** [comment]
EC2 requires that you obtain an IPv6 address via DHCPv6. While busybox ships with a client `updcpc6` it just seems to do the wrong thing pretty consistently. I was able to fix this by using the ISC `dhclient` client. It seems a little bit of a hack but this is what I did:

```
apk add dhclient

cat <<EOF >> /etc/network/interfaces

iface eth0 inet6 manual
  post-up dhclient -6 eth0
EOF

/etc/init.d/networking restart
```

There might be a better way to do this and if you find it please let me know. Otherwise I'll rename this issue and keep it open to track adding IPv6 DHCP support to our next release by default.

**Comment by tsarna** [comment]
Yes, I think this is a general Alpine issue and not AWS-specific. Sadly udhcpc6 seems fairly broken, pretty much unmaintained, and has near zero documentation as far as I can tell.

**Comment by mcrute** [comment]
That's the view I had too. Looking at the `ifup` soruce in busybox suggests it's not used at all there. I lack the knowledge or interest to make that work correctly for all the busybox use-cases (pretty much all my IPv6 use is SLAAC). So, unless there's a better way, the approach I outlined above will have to work.

**Comment by zfLQ2qx2** [comment]
@mcrute Thank you for the help! I found that I also had to do "echo 2 > /proc/sys/net/ipv6/conf/eth0/accept_ra" because Amazon is advertising an fe80::... address as the default route - though I don't know if that is the same for everyone or because I'm in a vpc private subnet behind an egress-only internet gateway.

**Comment by tomalok** [comment]
@mcrute @zfLQ2qx2 @tsarna 

As I'm doing a major overhaul of my personal Alpine AMI config, I took a look at this today...

I'm not sure it makes sense to set up DHCPv6 in EC2...  IPv6 addresses get assigned to interfaces, those addresses can be determined by querying the instance metadata service...  From one of my shell scripts...

``` shell
IMDS2='X-aws-ec2-metadata-token'

meta_token() {
  echo -ne "PUT /latest/api/token HTTP/1.0\r\n$IMDS2-ttl-seconds: 5\r\n\r\n" |
    nc 169.254.169.254 80 | tail -n 1
}

meta() {
  wget -qO - --header "$IMDS2: $(meta_token)" \
    "http://169.254.169.254/latest/meta-data/$1" 2>/dev/null
}

IPV6="$(meta network/interfaces/macs/02:87:a7:ef:03:5b/ipv6s)"

$ echo $IPVS
2600:1f13:224:d201:9542:f3ea:7366:6293
````

Maybe it'd make sense for `tiny-ec2-bootstrap` (or a different script) to look for these, and set up `/etc/network/interfaces` with static IPV6.

**Comment by zfLQ2qx2** [comment]
@tomalok I like that idea!

**Comment by tomalok** [comment]
I played with this a little in a timebox I set aside for myself yesterday -- I didn't have any luck with `udhcpc6`, `dhclient`, or `rdnssd` (SLAAC autodiscovery instead of DHCP) -- finding any helpful docs or understanding some of the finer points of configuring it properly is a bit of a challenge.

I did find some nice Ubuntu documentation about setting up IPv6 various ways with `/etc/network/interfaces`, but Alpine didn't seem to understand `iface eth0 inet6 auto` or `iface eth0 inet6 dhcp`...

From memory, some luck was had with extending `iface eth0 inet dhcp` with a `post-up` section that did a `ip addr add ...` and was able to SSH into the instance using its IPv6 address -- of course then I realized that I needed to setup `ip6tables` and sort out more security group rules and, well, I ran out of time in that box and needed to move along...

But a solution that would keep it small and easy would be for `tiny-ec2-bootstrap` to _optionally_ add a hook to run a small `post-up` script that checks the metadata service for any IPv6 for eth0's MAC and `ip addr add`s it.

**Comment by tomalok** [comment]
Opened up https://github.com/mcrute/tiny-ec2-bootstrap/pull/14 to allow optional auto-configuration of EC2 IPv6.

**Comment by mcrute** [comment]
@tomalok I tried `udhcpc6` and it seems to not work at all. Looking at the source for busybox it seems like it's just an incomplete implementation of DHCPv6.

As I recall [this is the document](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-migrate-ipv6.html#vpc-migrate-ipv6-dhcpv6) that I was referencing in the solution above. EC2 does send out RAs (see below) but they set the managed flag so SLAAC doesn't work but default route learning does. My guess is they do this because the control plane effectively assigns static IP addresses to the instance when it's launched. Also VPC doesn't support multicast so DAD doesn't work so there's the possibility for uncontrolled address generation collisions and pretty much invalidates SLAAC.

Given the way the control plane assigns IP addresses it seems a real waste to run 2 DHCP daemons just to read effectively static configuration. My main concern with this solution is that it's surprising for someone who is expecting to see the "standard" approach the AWS uses. I think a better approach would be to make the dhcp client within busybox pluggable so that ipv6 dhcp works in Alpine.

Here's an RA from our test VPC:
```
00:11:22.851365 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 56) fe80::8e8:5eff:fe81:925e > ff02::1: [icmp6 sum ok] ICMP6, router advertisement, length 56
        hop limit 64, Flags [managed], pref medium, router lifetime 1800s, reachable time 0ms, retrans timer 0ms
          source link-address option (1), length 8 (1): 0a:e8:5e:81:92:5e
            0x0000:  0ae8 5e81 925e
          prefix info option (3), length 32 (4): 2600:1f13:71a:6f02::/64, Flags [onlink], valid time infinity, pref. time infinity
            0x0000:  4080 ffff ffff ffff ffff 0000 0000 2600
            0x0010:  1f13 071a 6f02 0000 0000 0000 0000
```

**Comment by tomalok** [comment]
I have a `udhcpc` hook script that correctly syncs IP6s that are configured on the ENI at `post-bound` and `post-renew` events.  Doing the secondary IPv4s from the ENIs pretty much follows the same pattern.  I'm up in the air whether hooking into `pre-deconfig` to clean stuff up is necessary or not -- it seems to adversely impact `service networking restart`, so I'll probably skip that (will double check what Amazon Linux is doing).

**Comment by tomalok** [comment]
I think PR #85 is ready, and I've added a description of what's going on under the hood there.

**Comment by tomalok** [comment]
Closing this -- we now have IPv6 enabled by default, though _indirectly_ via IPv4 DHCP (`udhcpc` post-bound and post-renew hooks).

**Comment by tomalok** [comment]
We recently built and [released Alpine 3.12.1](https://github.com/mcrute/alpine-ec2-ami/blob/master/releases/README.md) AMIs a couple days ago.

These new images also fully support IMDSv2 and hotpluggable ENIs/IPv6/secondary IPv4s, and are also available for aarch64.
